services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/TransferProgressUpdaterTest.java

```diff
@@ -0,0 +1,186 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.ArgumentMatchers;
+import org.mockito.Mockito;
+import org.reactivestreams.Subscriber;
+import software.amazon.awssdk.core.SdkResponse;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.http.async.SimpleSubscriber;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.testutils.RandomTempFile;
+import software.amazon.awssdk.transfer.s3.CaptureTransferListener;
+import software.amazon.awssdk.transfer.s3.internal.progress.TransferProgressUpdater;
+import software.amazon.awssdk.transfer.s3.model.CompletedObjectTransfer;
+import software.amazon.awssdk.transfer.s3.model.TransferObjectRequest;
+import software.amazon.awssdk.transfer.s3.progress.LoggingTransferListener;
+import software.amazon.awssdk.transfer.s3.progress.TransferListener;
+
+class TransferProgressUpdaterTest {
+    private static final long OBJ_SIZE = 16 * MB;
+    private static File sourceFile;
+    private CaptureTransferListener captureTransferListener;
+
+    private static Stream<Arguments> contentLength() {
+
+        return Stream.of(
+            // When bytesTransfer ration reaches 1.0
+            Arguments.of(OBJ_SIZE),
+            // When bytesTransfer ration reaches  more than1.0
+            Arguments.of(OBJ_SIZE / 2),
+            // When mbytesTransfer ratio never reaches 100 , but subscription competes
+            Arguments.of(OBJ_SIZE * 2));
+    }
+
+    private static CompletableFuture<CompletedObjectTransfer> completedObjectResponse(long millis) {
+        return CompletableFuture.supplyAsync(() -> {
+            quietSleep(millis);
+            return new CompletedObjectTransfer() {
+                @Override
+                public SdkResponse response() {
+                    return PutObjectResponse.builder().eTag("ABCD").build();
+                }
+            };
+        });
+    }
+
+    private static void quietSleep(long millis) {
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt(); // Restore interrupted status
+            throw new RuntimeException(e);
+        }
+    }
+
+    @BeforeEach
+    void initiate() {
+        captureTransferListener = new CaptureTransferListener();
+    }
+
+    @ParameterizedTest
+    @MethodSource("contentLength")
+    void transferCompleteBasedOnBytesRatioWhenCompleted(Long adjustedContentLength) throws Exception {
```

Comment:
How does this test prove that the ratio 1 will complete the future? It takes content lengths that are both smaller and larger than the 1.0 ratio, and everything gets completed. I don't see where the difference is compared to before the change.

Reply:
There is a lot to unpack in this test case for the code reader not familiar with the code or the test. I suggest something like the following to improve readability:

~~~
    private static Stream<Arguments> contentLength() {
        return Stream.of(
           Arguments.of(OBJ_SIZE, "Total bytes equals transferred, future complete through subscriberOnNext()"),
           Arguments.of(OBJ_SIZE / 2, "Total bytes less than transferred, future complete through subscriberOnNext()"),
           Arguments.of(OBJ_SIZE * 2, "Total bytes more than transferred, future complete through subscriberOnComplete()"));
    }

@ParameterizedTest(name = "{index} - {1}, total bytes = {0}")
    @MethodSource("contentLength")
    void registerCompletion_differentTransferredByteRatios_alwaysCompletesOnce(Long givenContentLength, String description)
~~~

Which gives a test output of 
~~~
registerCompletion_differentTransferredByteRatios_alwaysCompletesOnce(Long, String)
1 - Total bytes equals transferred, future complete through subscriberOnNext(), total bytes = 16777216
2 - Total bytes less than transferred, future complete through subscriberOnNext(), total bytes = 8388608
3 - Total bytes more than transferred, future complete through subscriberOnComplete(), total bytes = 33554432
~~~
(it may not even be meaningful to output the exact bytes)

- The test name indicates that the complete once is the important metric for the test
- The test case descriptions shows that there is a difference between the two cases

Reply:
Thanks